lib/mixins/dir-mixin.html

```diff
@@ -0,0 +1,130 @@
+  (function() {
+    'use strict';
+
+    const HOST_DIR = /:host\(:dir\((ltr|rtl)\)\)/g;
+    const HOST_DIR_REPLACMENT = ':host([dir="$1"])';
+
+    const EL_DIR = /([\s\w#\.\[\]\*]*):dir\((ltr|rtl)\)/g;
+    const EL_DIR_REPLACMENT = ':host([dir="$2"]) $1';
+
+    /**
+     * @type {!Array<!Polymer_DirMixin>}
+     */
+    const DIR_INSTANCES = [];
+
+    let observed = false;
+
+    /**
+     * @param {!Polymer_DirMixin} instance Instance to set RTL status on
+     */
+    function setRTL(instance) {
+      if (!instance.__origRTLStatus) {
+        const el = /** @type {!HTMLElement} */(instance);
+        if (document.documentElement.getAttribute('dir') === 'rtl') {
+          el.setAttribute('dir', 'rtl');
+        } else {
+          el.removeAttribute('dir');
+        }
+      }
+    }
+
+    /**
+     * Element class mixin that allows elements to use the `:dir` CSS Selector to have
+     * text direction specific styling.
+     *
+     * With this mixin, any stylesheet provided in the template will be transform `:dir` into
+     * `:host([dir])` and sync direction with the page via the element's `dir` attribute.
+     *
+     * Elements can opt out of the global page text direction by setting the `dir` attribute
+     * directly in `ready()`.
+     *
+     * Caveats:
+     * - Automatic left-to-right or right-to-left styling is sync'd with the `<html>` element
+     * - Applications must set `<html dir="ltr">` or `<html dir="rtl">` to sync direction
+     * - Opting out of the global direction styling is permanent
+     *
+     * @mixinFunction
+     * @polymer
+     * @appliesMixin Polymer.ElementMixin
+     * @memberof Polymer
+     */
+    Polymer.DirMixin = Polymer.dedupingMixin((base) => {
+
+      if (!observed) {
+        new MutationObserver(() => {
+          DIR_INSTANCES.forEach(i => setRTL(i));
+        }).observe(document.documentElement, {attributes: true, attributeFilter: ['dir']});
+      }
+
+      /**
+       * @constructor
+       * @extends {base}
+       * @implements {Polymer_ElementMixin}
+       */
+      const elementBase = Polymer.ElementMixin(base);
+
+      /**
+       * @polymer
+       * @mixinClass
+       * @implements {Polymer_DirMixin}
+       */
+      class Dir extends elementBase {
+
+        /**
+         * @override
+         * @suppress {missingProperties} Interfaces in closure do not inherit statics, but classes do
+         */
+        static _processStyleText(is, template, baseURI) {
+          let cssText = super._processStyleText(is, template, baseURI);
+          cssText = cssText.replace(HOST_DIR, HOST_DIR_REPLACMENT);
+          cssText = cssText.replace(EL_DIR, EL_DIR_REPLACMENT);
+          return cssText;
+        }
+
+        constructor() {
+          super();
+          /** @type {boolean} */
+          this.__origRTLStatus = false;
+        }
+
+        /**
+         * @suppress {invalidCasts} Closure doesn't understand that `this` is an HTMLElement
+         */
+        ready() {
```

Comment:
Let's make these runtime features opt-in via a flag.

Reply:
Not sure if this is still needed if we use the presence of `:dir` to activate.

Reply:
Seems ok. They are different and an element could have either or both.